File: solutions/Wildcard-Matching/Wildcard-Matching.py


class Solution:
    # @param s, an input string
    # @param p, a pattern string
    # @return a boolean
    def isMatch(self, s, p):
        len_s = len(s); len_p = len(p)
        pPointer = sPointer = ss = 0
        star = -1
        while sPointer < len_s:
            if pPointer < len_p and (p[pPointer] == '?' or p[pPointer] == s[sPointer]):
                sPointer += 1; pPointer += 1;
                continue
            if pPointer < len_p and p[pPointer] == '*':
                star = pPointer; pPointer +=1; ss = sPointer
                continue
            if star != -1:
                pPointer = star + 1; ss += 1; sPointer = ss
                continue
            return False
        while pPointer < len_p and p[pPointer] == '*':
            pPointer += 1
        return pPointer == len_p
 

File: solutions/Wildcard-Matching/Wildcard-Matching.js
/**
 * @param {string} s
 * @param {string} p
 * @return {boolean}
 */
var isMatch = function (s, p) {
  const n = s.length;
  const m = p.length;
  let star = -1;
  let prev = -1;
  let j = 0;
  for (let i = 0; i < n;) {
    if (s[i] === p[j] || p[j] === '?') {
      i++;
      j++;
      continue;
    }
    if (p[j] === '*') {
      star = j;
      j++;
      prev = i;
      continue;
    }
    if (star !== -1) {
      j = star + 1;
      i = prev + 1;
      prev++;
      continue;
    }
    return false;
  }
  while (p[j] === '*') {
    j++;
  }
  return j === m;
};